### PR TITLE
#1040 SMILES/SMARTS import: Files with S-Group cause an error 

### PR DESCRIPTION
**Pull request by AliaksandrDziarkach**
Implemented import support for Data S-Groups.
 Implemented import support for all fields for Polymer S-Groups.
 Implemented export support for Data S-Groups.
 Implemented export support for all fields for Polymer S-Groups.
 Implemented corresponding UTs into indigo-core-unit-tests.
 Fix integration UTs.

### CODE REVIEW

File: core/indigo-core/molecule/src/smiles_loader.cpp [hunk 1]
```diff
@@ -76,6 +76,30 @@ void SmilesLoader::loadQueryMolecule(QueryMolecule& mol)
...
 
+static char readSgChar(Scanner& scanner)
+{
+    char c = scanner.readChar();
+    if (c == '&' && scanner.lookNext() == '#') // Escaped char - &#code;
+    {
+        long long pos = scanner.tell();
+        scanner.skip(1); // Skip '#'
+        int code = scanner.tryReadUnsigned();
+        if (code >= 0 && code < 256 && scanner.lookNext() == ';')
```
Comment (AlexanderSavelyev) [inline core/indigo-core/molecule/src/smiles_loader.cpp:87]: please move the 256 to a constant

File: core/indigo-core/molecule/src/smiles_loader.cpp [hunk 2]
```diff
@@ -860,28 +908,165 @@ void SmilesLoader::_readOtherStuff()
...
+                }
+                if (_scanner.lookNext() != ':') // No more fields
+                    continue;
+                _scanner.skip(1); // Skip :
+                // (coords)
+                c = _scanner.lookNext();
+                if (c != '(') // No more fields
+                    continue;
+                long long pos = _scanner.tell();
+                if (_scanner.length() - pos >= 4)
```
Comment (AlexanderSavelyev) [inline core/indigo-core/molecule/src/smiles_loader.cpp:956]: please move 4 to the constant

File: core/indigo-core/molecule/src/smiles_loader.cpp [hunk 3]
```diff
@@ -860,28 +908,165 @@ void SmilesLoader::_readOtherStuff()
...
+                {
+                    // check for (-1)
+                    char buf[5];
```
Comment (AlexanderSavelyev) [inline core/indigo-core/molecule/src/smiles_loader.cpp:959]: 5 to the constant

File: core/indigo-core/molecule/src/smiles_loader.cpp [hunk 4]
```diff
@@ -808,35 +832,61 @@ void SmilesLoader::_readOtherStuff()
...
+            // *_bond_indexes - The indexes of bonds that share a common bracket in case of ladder-type polymers.
+            // head_bond_indexes - Head crossing bond indexes. This field can be empty.
+            // tail_bond_indexes - Tail crossing bond indexes. This field can be empty.
+            // bracket - bracket orientation, bracket type followed by the coordinates (4 pair, separated with commas). Bracket orientation
+            //     can be s or d (single or double), bracket type can be b,c,r,s for braces, chevrons, round and square, respectively.
+            //     The brackets are written between parentheses and separated with semicolons.
+            if (sg_type == -1)
+            {
+                char sg = _scanner.lookNext();
+                char pchar_sg_type[3];
```
Comment (AlexanderSavelyev) [inline core/indigo-core/molecule/src/smiles_loader.cpp:864]: add constant for 3 (is it max sgroup name len?)

File: core/indigo-core/molecule/src/smiles_loader.cpp [hunk 5]
```diff
@@ -860,28 +908,165 @@ void SmilesLoader::_readOtherStuff()
...
+                char br_type = _scanner.readChar();
+                c = _scanner.readChar();
+                int count = 0;
+                while (c == ',' && count < 8)
+                {
+                    float tmp = _scanner.readFloat();
+                    c = _scanner.readChar();
+                    ++count;
+                }
+                if (count < 8)
```
Comment (AlexanderSavelyev) [inline core/indigo-core/molecule/src/smiles_loader.cpp:1064]: move 8 to constants

**Review by AlexanderSavelyev — CHANGES_REQUESTED** [review]

**Review by even1024 — APPROVED** [review]

**Pull request merged**